### ISSUE TITLE
CSRF Setup with beaker

### ISSUE THREAD

**Issue by AvnerCohen** [Pylons/pyramid#3170]
Thanks for all the work Pyramid !

I'm really confused about the setup to fully support CSRF in an app.
From XHR to form hidden fields, session type cookie and beaker.
I tried to find any reference I could find, but so far failed.

I tried to follow the basic setup as per:
https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/security.html#checking-csrf-tokens-automatically
Basically:
```
    config.set_default_csrf_options(require_csrf=True)
```
and than, in my mako template:
```
          <input type="hidden" name="csrf_token" value="${get_csrf_token()}">
```
However, regardless of the setup I tested, I couldn't get this to work.

I suspect there is some conflict when I setup:
    config.set_csrf_storage_policy(SessionCSRFStoragePolicy())    
    config.include("pyramid_beaker")

Because I see crft token being setup here:
https://github.com/Pylons/pyramid_beaker/blob/master/pyramid_beaker/__init__.py#L90
But also here:
https://github.com/Pylons/pyramid/blob/master/pyramid/csrf.py#L80

When I print out the tokens on each post, I see that when I try to manually run:
```
from pyramid.csrf import check_csrf_token

print "GOT: %s" % request.params.get('csrf_token')
print "FROM SESSION: %s" % request.session.get_csrf_token()
print "Inside check %s" % check_csrf_token(request)
````

The form sends the token from SessionCSRFStoragePolicy, but get_csrf_token returns the one from pyramid_beaker (end as a result check_csrf_token fails with BadCSRFToken.

Any thoughts?
Any additional info I should be providing?

**Comment by mmerickel** [comment]
SessionCSRFStoragePolicy != LegacySessionCSRFStoragePolicy

Basically if you want get_csrf_token to match request.session.get_csrf_token then you must use the legacy policy.

**Comment by AvnerCohen** [comment]
Thank you @mmerickel I tried the same with LegacySessionCSRFStoragePolicy , but still unable to get this to work.

My setup is as above, simply changed:
`    config.set_csrf_storage_policy(SessionCSRFStoragePolicy())    `
to:
`     config.set_csrf_storage_policy(LegacySessionCSRFStoragePolicy())   `

When I manually invoke `check_csrf_token()` I get a csrf failure.

But this is just part of the test, what I am trying to setup is the simplest *Checking CSRF Tokens Automatically* , nothing fancy.

When I do not add the manual check, csrf is never failing (even If I manually send a bad token), which makes me think `check_csrf_token` is no being called at all, despite:
`config.set_default_csrf_options(require_csrf=True)`

Would love to help with the docs when I get this working, any additional reference I can work with to try and better what is it that I'm doing wrong?

**Comment by AvnerCohen** [comment]
Further debugging on this suggests that this might be related to a forbidden_view_config that does not invoke `csrf_view` in the same way.

I'm quite sure it's me doing something wrong, not yet sure what exactly though.

**Comment by AvnerCohen** [comment]
Found the issue. 100% bad setup on my part. thanks!

**Comment by rustybox** [comment]
Always good to describe your problem & solution in case this ever shows up on the googles.

**Comment by AvnerCohen** [comment]
:)
Roughly speaking, as there were multiple issues, I suspect my problem was a duplicated session manager setup that created 2 tokens with each POST requests, so the csrf tests failed and the overall behavior was just incosistent .

I couldn't 100% find out why and how, however the "fix" in my case was removing beaker (as it seems to take over csrf setting in some cases) and the code changes are:
```
     config = Configurator()
     config.set_default_csrf_options(require_csrf=True)
     config.set_csrf_storage_policy(LegacySessionCSRFStoragePolicy())
     config.include("pyramid_beaker")
```
To:
```
     config = Configurator()
     config.set_default_csrf_options(require_csrf=True)
     config.set_csrf_storage_policy(LegacySessionCSRFStoragePolicy())
     signed_session_factory = SignedCookieSessionFactory('super_secret_stuff')
     config.set_session_factory(signed_session_factory)
```

Basically, dropping beaker in favor of pyramid sessions in an attempt to reduce complexity..
